File: packages/tools/src/cli/commands/sign.ts
import { rm, writeFile } from "node:fs/promises";
import { join } from "node:path";
import { buildCommand, buildRouteMap } from "@stricli/core";
import { execa } from "execa";
import { getRepoRoot } from "../../lib/git.ts";

async function withRetry<T>(
  fn: () => Promise<T>,
  maxRetries = 2,
  delayMs = 30_000,
): Promise<T> {
  for (let attempt = 1; attempt <= maxRetries; attempt++) {
    try {
      return await fn();
    } catch (error) {
      if (attempt === maxRetries) throw error;
      console.log(
        `Attempt ${attempt}/${maxRetries} failed, retrying in ${delayMs / 1000}s...`,
      );
      await new Promise((resolve) => setTimeout(resolve, delayMs));
    }
  }
  throw new Error("Unreachable");
}

const signMacosCommand = buildCommand<
  { executablePath: string; skipNotarize: boolean },
  []
>({
  parameters: {
    flags: {
      executablePath: {
        kind: "parsed",
        brief: "Path to the executable to sign",
        parse: String,
      },
      skipNotarize: {
        kind: "boolean",
        brief: "Skip notarization (codesign only)",
        default: false,
      },
    },
  },
  docs: {
    brief: "Sign and Notarize macOS binary",
  },
  async func(flags) {
    const repoRoot = await getRepoRoot(process.cwd());
    const executablePath = join(repoRoot, flags.executablePath);
    const entitlementsPath = join(repoRoot, "packages/cli/entitlements.plist");

    // biome-ignore lint/complexity/useLiteralKeys: must use bracket notation for index signature access
    const appleCertificate = process.env["APPLE_CERTIFICATE"];
    // biome-ignore lint/complexity/useLiteralKeys: must use bracket notation for index signature access
    const appleCertificatePassword = process.env["APPLE_CERTIFICATE_PASSWORD"];
    // biome-ignore lint/complexity/useLiteralKeys: must use bracket notation for index signature access
    const appleDeveloperId = process.env["APPLE_DEVELOPER_ID"];
    // biome-ignore lint/complexity/useLiteralKeys: must use bracket notation for index signature access
    const appleNotaryKeyId = process.env["APPLE_NOTARY_KEY_ID"];
    // biome-ignore lint/complexity/useLiteralKeys: must use bracket notation for index signature access
    const appleNotaryIssuerId = process.env["APPLE_NOTARY_ISSUER_ID"];
    // biome-ignore lint/complexity/useLiteralKeys: must use bracket notation for index signature access
    const appleNotaryKeyP8Base64 = process.env["APPLE_NOTARY_KEY_P8_BASE64"];

    console.log("Removing existing signature if any...");
    try {
      await execa("codesign", ["--remove-signature", executablePath]);
    } catch {
      // Ignore if it fails (e.g. no signature exists)
    }

    console.log("Removing extended attributes if any...");
    try {
      await execa("xattr", ["-cr", executablePath]);
    } catch {
      // Ignore if it fails
    }

    if (appleCertificate) {
      if (!appleCertificatePassword || !appleDeveloperId) {
        throw new Error(
          "APPLE_CERTIFICATE_PASSWORD and APPLE_DEVELOPER_ID are required when APPLE_CERTIFICATE is set",
        );
      }

      console.log("Importing Apple Certificate...");
      const certBuffer = Buffer.from(appleCertificate, "base64");
      await writeFile("certificate.p12", certBuffer);

      try {
        try {
          await execa("security", ["delete-keychain", "build.keychain"]);
        } catch {
          // Ignore if keychain doesn't exist
        }
        await execa("security", [
          "create-keychain",
          "-p",
          "actions",
          "build.keychain",
        ]);
        const { stdout: currentKeychains } = await execa("security", [
          "list-keychains",
        ]);
        const keychainList = currentKeychains
          .split("\n")
          .map((k) => k.trim())
          .filter(Boolean);
        if (!keychainList.includes("build.keychain")) {
          await execa("security", [
            "list-keychains",
            "-s",
            ...keychainList,
            "build.keychain",
          ]);
        }
        await execa("security", ["default-keychain", "-s", "build.keychain"]);
        await execa("security", [
          "unlock-keychain",
          "-p",
          "actions",
          "build.keychain",
        ]);
        await execa("security", [
          "import",
          "certificate.p12",
          "-k",
          "build.keychain",
          "-P",
          appleCertificatePassword,
          "-T",
          "/usr/bin/codesign",
        ]);
        await execa("security", [
          "set-key-partition-list",
          "-S",
          "apple-tool:,apple:,codesign:",
          "-s",
          "-k",
          "actions",
          "build.keychain",
        ]);

        console.log("Signing macOS binary...");
        await execa("codesign", [
          "--force",
          "--options",
          "runtime",
          "--entitlements",
          entitlementsPath,
          "--sign",
          appleDeveloperId,
          executablePath,
        ]);

        if (appleNotaryKeyP8Base64 && !flags.skipNotarize) {
          if (!appleNotaryKeyId || !appleNotaryIssuerId) {
            throw new Error(
              "APPLE_NOTARY_KEY_ID and APPLE_NOTARY_ISSUER_ID are required when APPLE_NOTARY_KEY_P8_BASE64 is set",
            );
          }

          console.log("Notarizing macOS binary...");
          const authKeyBuffer = Buffer.from(appleNotaryKeyP8Base64, "base64");
          await writeFile("AuthKey.p8", authKeyBuffer);

          const zipPath = `${executablePath}.zip`;
          await execa("zip", ["-j", zipPath, executablePath]);

          await withRetry(() =>
            execa("xcrun", [
              "notarytool",
              "submit",
              zipPath,
              "--key",
              "AuthKey.p8",
              "--key-id",
              appleNotaryKeyId,
              "--issuer",
              appleNotaryIssuerId,
              "--wait",
            ]),
          );
          await rm("AuthKey.p8");
        } else if (flags.skipNotarize) {
          console.log("Notarization skipped (--skip-notarize flag).");
        } else {
          console.log("No Notary API Key found. Skipping notarization.");
        }
      } finally {
        await rm("certificate.p12", { force: true });
      }
    } else {
      console.log("No Apple Certificate found. Performing ad-hoc signing...");
      await execa("codesign", ["--sign", "-", executablePath]);
    }
  },
});

export const signRoute = buildRouteMap({
  routes: {
    macos: signMacosCommand,
  },
  docs: {
    brief: "Signing commands",
  },
});

File: packages/tools/lib/src/lib/sign.dart
import 'dart:convert';
import 'dart:io';

import 'package:path/path.dart' as p;

import 'error.dart';
import 'exec.dart';

Future<T> withRetry<T>(
  Future<T> Function() fn, {
  int maxRetries = 2,
  Duration delay = const Duration(seconds: 30),
}) async {
  for (var attempt = 1; attempt <= maxRetries; attempt++) {
    try {
      return await fn();
    } catch (_) {
      if (attempt == maxRetries) {
        rethrow;
      }

      stdout.writeln(
        'Attempt $attempt/$maxRetries failed, '
        'retrying in ${delay.inSeconds}s...',
      );
      await Future<void>.delayed(delay);
    }
  }

  throw StateError('Unreachable');
}

Future<void> _tryRun(String executable, List<String> args) async {
  try {
    await runChecked(executable, args);
  } on ToolException {
    // Ignore failures (e.g. no signature or attributes exist).
  }
}

/// Decodes a base64 secret leniently, like Node's `Buffer.from(s, 'base64')`
/// which the pre-cutover TS signing flow used: CI secrets produced with the
/// `base64` CLI wrap at 64/76 columns, and Dart's strict decoder rejects the
/// embedded newlines ("Invalid padding character").
List<int> decodeBase64Secret(String value) {
  return base64Decode(value.replaceAll(RegExp(r'\s'), ''));
}

Future<void> _deleteIfExists(String path) async {
  final file = File(path);

  if (file.existsSync()) {
    await file.delete();
  }
}

Future<void> performMacosSign({
  required String repoRoot,
  required String executablePath,
  required bool skipNotarize,
}) async {
  final resolvedExecutablePath = p.join(repoRoot, executablePath);
  final entitlementsPath = p.join(repoRoot, 'packages/cli/entitlements.plist');

  final env = Platform.environment;
  final appleCertificate = env['APPLE_CERTIFICATE'];
  final appleCertificatePassword = env['APPLE_CERTIFICATE_PASSWORD'];
  final appleDeveloperId = env['APPLE_DEVELOPER_ID'];
  final appleNotaryKeyId = env['APPLE_NOTARY_KEY_ID'];
  final appleNotaryIssuerId = env['APPLE_NOTARY_ISSUER_ID'];
  final appleNotaryKeyP8Base64 = env['APPLE_NOTARY_KEY_P8_BASE64'];

  stdout.writeln('Removing existing signature if any...');
  await _tryRun('codesign', ['--remove-signature', resolvedExecutablePath]);

  stdout.writeln('Removing extended attributes if any...');
  await _tryRun('xattr', ['-cr', resolvedExecutablePath]);

  if (appleCertificate != null && appleCertificate.isNotEmpty) {
    if (appleCertificatePassword == null ||
        appleCertificatePassword.isEmpty ||
        appleDeveloperId == null ||
        appleDeveloperId.isEmpty) {
      throw const ToolException(
        'APPLE_CERTIFICATE_PASSWORD and APPLE_DEVELOPER_ID are required '
        'when APPLE_CERTIFICATE is set',
      );
    }

    stdout.writeln('Importing Apple Certificate...');
    await File(
      'certificate.p12',
    ).writeAsBytes(decodeBase64Secret(appleCertificate));

    try {
      await _tryRun('security', ['delete-keychain', 'build.keychain']);
      await runChecked('security', [
        'create-keychain',
        '-p',
        'actions',
        'build.keychain',
      ]);

      final listResult = await runChecked('security', ['list-keychains']);
      final keychainList = listResult.stdout
          .split('\n')
          .map((keychain) => keychain.trim())
          .where((keychain) => keychain.isNotEmpty)
          .toList();

      if (!keychainList.contains('build.keychain')) {
        await runChecked('security', [
          'list-keychains',
          '-s',
          ...keychainList,
          'build.keychain',
        ]);
      }

      await runChecked('security', [
        'default-keychain',
        '-s',
        'build.keychain',
      ]);
      await runChecked('security', [
        'unlock-keychain',
        '-p',
        'actions',
        'build.keychain',
      ]);
      await runChecked('security', [
        'import',
        'certificate.p12',
        '-k',
        'build.keychain',
        '-P',
        appleCertificatePassword,
        '-T',
        '/usr/bin/codesign',
      ]);
      await runChecked('security', [
        'set-key-partition-list',
        '-S',
        'apple-tool:,apple:,codesign:',
        '-s',
        '-k',
        'actions',
        'build.keychain',
      ]);

      stdout.writeln('Signing macOS binary...');
      await runChecked('codesign', [
        '--force',
        '--options',
        'runtime',
        '--entitlements',
        entitlementsPath,
        '--sign',
        appleDeveloperId,
        resolvedExecutablePath,
      ]);

      if (appleNotaryKeyP8Base64 != null &&
          appleNotaryKeyP8Base64.isNotEmpty &&
          !skipNotarize) {
        if (appleNotaryKeyId == null ||
            appleNotaryKeyId.isEmpty ||
            appleNotaryIssuerId == null ||
            appleNotaryIssuerId.isEmpty) {
          throw const ToolException(
            'APPLE_NOTARY_KEY_ID and APPLE_NOTARY_ISSUER_ID are required '
            'when APPLE_NOTARY_KEY_P8_BASE64 is set',
          );
        }

        stdout.writeln('Notarizing macOS binary...');
        await File(
          'AuthKey.p8',
        ).writeAsBytes(decodeBase64Secret(appleNotaryKeyP8Base64));

        final zipPath = '$resolvedExecutablePath.zip';
        await runChecked('zip', ['-j', zipPath, resolvedExecutablePath]);

        await withRetry(
          () => runChecked('xcrun', [
            'notarytool',
            'submit',
            zipPath,
            '--key',
            'AuthKey.p8',
            '--key-id',
            appleNotaryKeyId,
            '--issuer',
            appleNotaryIssuerId,
            '--wait',
          ]),
        );
        await _deleteIfExists('AuthKey.p8');
      } else if (skipNotarize) {
        stdout.writeln('Notarization skipped (--skip-notarize flag).');
      } else {
        stdout.writeln('No Notary API Key found. Skipping notarization.');
      }
    } finally {
      await _deleteIfExists('certificate.p12');
    }
  } else {
    stdout.writeln('No Apple Certificate found. Performing ad-hoc signing...');
    await runChecked('codesign', ['--sign', '-', resolvedExecutablePath]);
  }
}
